clarity/src/vm/analysis/type_checker/contexts.rs

```diff
@@ -45,7 +47,11 @@ pub struct ContractContext {
     persisted_variable_types: HashMap<ClarityName, TypeSignature>,
     fungible_tokens: HashSet<ClarityName>,
     non_fungible_tokens: HashMap<ClarityName, TypeSignature>,
+    // Before Clarity2, traits are keyed by the trait name.
     traits: HashMap<ClarityName, BTreeMap<ClarityName, FunctionSignature>>,
+    // In Clarity2+, we key by the whole trait identifier, and keep track of defined traits.
+    defined_traits: HashSet<ClarityName>,
+    all_traits: HashMap<TraitIdentifier, BTreeMap<ClarityName, FunctionSignature>>,
```

Comment:
I'm not sure I understand the differences between `traits`, `defined_traits`, and `all_traits`?

Reply:
`traits` is kept to preserve the Clarity1 behavior, where used traits are treated the same as defined traits. `all_traits` is a place to store all of the traits which are defined or used in a contract, and `defined_traits` is making it easy to lookup locally defined traits by their alias.

I will review again to see if this can be simplified now.

Reply:
Can you add this information to `///` rustdoc comments on these fields?

Reply:
Actually, looking at how this gets used, do you think it makes the struct simpler to use like an enum here?

```rust
enum TraitContext {
    /// Traits stored in this context use the trait type-checking behavior defined in Clarity1
    Clarity1(HashMap),
    /// Traits stored in this context use the new trait type-checking behavior defined in Clarity2
    Clarity2 {
         /// Aliases for locally defined traits
         defined: HashSet,
         /// All traits which are defined or used in a contract
         all: HashMap
    },
}

pub struct ContractContext {
    ..
    traits: TraitContext
}
```

Then, instead of the version checks in the code, we could match on that field?

E.g.,

```rust
 pub fn add_defined_trait(
        &mut self,
        trait_name: ClarityName,
        trait_signature: BTreeMap<ClarityName, FunctionSignature>,
    ) -> CheckResult<()> {
        match self.traits {
             Clarity1(ref mut traits) => {
                 traits.insert(trait_name, trait_signature);
             }
             ...
        }
        Ok(())
    }
```

I'm not sure if that makes things actually more clear, but it would capture the kind of either-or nature of `traits` and (`defined_traits`, `all_traits`).

Reply:
I agree with @kantai.  The use of an `enum` would prevent us from accidentally mutating Clarity2-specific state from Clarity1 code paths, which I'm keen to guard against in this code to the greatest extent possible.

Reply:
good idea.. is this done?

Reply:
It is now. Implemented in edec454c2.